src/test/rgw/test_cls_fifo_legacy.cc

```diff
@@ -133,6 +133,87 @@ TEST_F(LegacyClsFIFO, TestGetInfo)
   EXPECT_EQ(-ECANCELED, r);
 }
 
+TEST_F(LegacyClsFIFO, TestDoubleJournalRefusal)
```

Comment:
So, we are just changing the unittest here but create_part is still non-exclusive, right? Again from my testing on top of your Nov 1 head (https://github.com/ceph/ceph/commit/cfc3bde36dbc9c6e0b7182bbb32539019ba2dd61), added below two relevant commits in my tests:

https://github.com/adamemerson/ceph/pull/3/commits/1809444e4441affaaa52f4c1ae49d31d0f5a9aa4 (let process journal return the true error value without negating) and 

https://github.com/adamemerson/ceph/pull/3/commits/72023ed0e481dba4414214e5d41bba2f4f48a008 (making create_part exclusive)

Since create_op & set_head come in pairs, this would disallow a set_head if create_part fails with -EEXIST. 

Do the latest changes block a second set_head to execute for a given part_num? If so, making this exlcusive wouldn't be needed.

Another relevant commit https://github.com/adamemerson/ceph/pull/3/commits/216c6e0696b156b14c48aadab90508ad8cc0ffd6 removes stale set_heads rather than letting them blocking subsequent set_heads for new part_nums.

I noticed a run in which https://github.com/ceph/ceph/blob/74848f8b0a7b4fa13e286351bb7da91d11b8cbaa/src/rgw/cls_fifo_legacy.cc#L793 causes stale ops remain in the journal and the next create_part is blocked as a result. E.g., thread 1 adds the ops to journal and start processing the tmp_journal and removes those ops. Soon, later thread2 adds ops but sees things are all good ("raced, journaled but not processed") and moves one leaving stale ops in the journal.

Again, perhaps the new changes address this race condition and addressing stale create_op/set_heads in the journal is not issue anymore.

Reply:
This wasn't changing the unit test, I just wanted to verify that the safeguards I thought were in place are actually in place.